src/arrays/quarray.jl

```diff
@@ -0,0 +1,143 @@
+import Base: size,
+    length,
+    getindex,
+    similar,
+    in,
+    ctranspose,
+    transpose,
+    summary,
+    zeros,
+    eye,
+    #TODO: Implement the below operations
+    *,.*,
+    /,./,
+    +,.+,
+    -,.-,
+    kron
+
+abstract AbstractQuArray{B<:AbstractBasis,T,N} <: AbstractArray{T,N}
+
+###########
+# QuArray #                 
+###########
+    # Using an NTuple allows us to have a basis for each dimension, 
+    # and gives us a less ambiguous way to determine if a QuArray will act
+    # like a vector, matrix, or tensor using the dimension parameter N. 
+    type QuArray{B<:AbstractBasis,T,N,A} <: AbstractQuArray{B,T,N}
+        coeffs::A
+        bases::NTuple{N,B}
+        function QuArray(coeffs::AbstractArray{T}, bases::NTuple{N,B}) 
+            if checkbases(coeffs, bases) 
+                new(coeffs, bases)
+            else 
+                error("Coefficient array does not conform to input bases")
+            end
+        end
+    end
+    
+    typealias QuVector{B<:AbstractBasis,T,A} QuArray{B,T,1,A}
+    typealias QuMatrix{B<:AbstractBasis,T,A} QuArray{B,T,2,A}
+
+    QuArray{T,N,B<:AbstractBasis}(coeffs::AbstractArray{T}, bases::NTuple{N,B}) = QuArray{B,T,N,typeof(coeffs)}(coeffs, bases)
+    QuArray(coeffs, bases::AbstractBasis...) = QuArray(coeffs, bases)
+    QuArray(coeffs) = QuArray(coeffs, basesfordims(size(coeffs)))
+    
+    ############################
+    # Convenience Constructors #
+    ############################
+    statevec(i::Int, fb::FiniteBasis) = QuArray(single_coeff(i, length(fb)), fb)
+    statevec(i::Int, lens::Int...=i) = statevec(i, FiniteBasis(lens))
+
+    ######################
+    # Property Functions #
+    ######################
+    bases(qa::QuArray) = qa.bases
+    coeffs(qa::QuArray) = qa.coeffs
+
+    size(qa::QuArray, i...) = size(qa.coeffs, i...)
+
+    ########################
+    # Array-like functions #
+    ########################
+    similar{B,T}(qa::QuArray{B,T}, element_type=T) = QuArray(similar(qa.coeffs, T), qa.bases)
+    # is there a way to properly define the below for
+    # any arbitrary basis? Obviously doesn't make sense
+    # for B<:AbstractInfiniteBasis, and I'm reluctant to
+    # enforce that every B<:AbstractFiniteBasis will have a 
+    # constructor B(::Int), which is how the below is constructing
+    # instances of FiniteBasis.
+    function similar{B<:FiniteBasis,T}(qa::QuArray{B,T}, element_type, dims)
+        return QuArray(similar(qa.coeffs, T, dims), basesfordims(dims, B))
+    end 
+    getindex(qa::QuArray, i::AbstractArray) = getindex(qa.coeffs, i)
+    getindex(qa::QuArray, i::Real) = getindex(qa.coeffs, i)
+    getindex(qa::QuArray, i) = getindex(qa.coeffs, i)
+    getindex(qa::QuArray, i...) = getindex(qa.coeffs, i...)
+
+    in(c, qa::QuArray) = in(c, qa.coeffs)
+
+    ctranspose(qa::QuVector) = QuArray(ctranspose(qa.coeffs), qa.bases)
+    ctranspose(qa::QuMatrix) = QuArray(ctranspose(qa.coeffs), reverse(qa.bases))
```

Comment:
I like the fact that a transposed `QuVector` stays a `QuVector` and doesn't become a matrix like for arrays. But we might want to have some way to indicate that a `QuVector` or a `QuMatrix` was transposed. I am thinking of something like JuliaLang/julia#6837, i.e. introducing a dual/transposed type. This would make it easier to define inner products and such. We don't have to do this now. Maybe once this is merged we can have an issue as a reminder. cc: @Jutho